src/legacy/ui/public/agg_types/controls/time_interval.tsx

```diff
@@ -0,0 +1,172 @@
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n/react';
+import { AggParamEditorProps } from '../../vis/editors/default';
+import { AggParamOption } from '../agg_param';
+import { parseInterval } from '../utils';
+
+interface ComboBoxOption extends EuiComboBoxOptionProps {
+  key: string;
+}
+
+function TimeIntervalParamEditor({
+  agg,
+  aggParam,
+  editorConfig,
+  value,
+  setValue,
+  isInvalid,
+  setTouched,
+  setValidity,
+}: AggParamEditorProps<string>) {
+  const interval = get(agg, 'buckets.getInterval') && agg.buckets.getInterval();
+  const scaledHepText =
+    interval && interval.scaled && !isInvalid ? (
+      <strong className="eui-displayBlock">
+        <FormattedMessage
+          id="common.ui.aggTypes.timeInterval.scaledHelpText"
+          defaultMessage="Currently scaled to {bucketDescription}"
+          values={{ bucketDescription: get(interval, 'description') || '' }}
+        />{' '}
+        <EuiIconTip
+          position="right"
+          type="questionInCircle"
+          content={interval.scale <= 1 ? tooManyBucketsTooltip : tooLargeBucketsTooltip}
+        />
+      </strong>
+    ) : null;
+
+  const helpText = (
+    <>
+      {scaledHepText}
+      {get(editorConfig, 'interval.help') || selectOptionHelpText}
+    </>
+  );
+
+  const timeBase: string = get(editorConfig, 'interval.timeBase');
+  const options = timeBase
+    ? []
+    : (aggParam.options || []).reduce(
+        (filtered: ComboBoxOption[], option: AggParamOption) => {
+          if (option.enabled ? option.enabled(agg) : true) {
+            filtered.push({ label: option.display, key: option.val });
+          }
+          return filtered;
+        },
+        [] as ComboBoxOption[]
+      );
+
+  let selectedOptions: ComboBoxOption[] = [];
+  let definedOption: ComboBoxOption | undefined;
+  if (value) {
+    definedOption = find(options, { key: value });
+    selectedOptions = definedOption ? [definedOption] : [{ label: value, key: 'custom' }];
+  }
+
+  const errors = [];
+
+  if (isInvalid && value) {
+    errors.push(
+      i18n.translate('common.ui.aggTypes.timeInterval.invalidFormatErrorMessage', {
+        defaultMessage: 'Invalid interval format.',
+      })
+    );
+  }
+
+  const onCustomInterval = (customValue: string) => {
+    const normalizedCustomValue = customValue.trim();
+    const isValid = normalizedCustomValue ? parseInterval(normalizedCustomValue, timeBase) : false;
+
+    setValidity(isValid);
+    setValue(normalizedCustomValue);
+    setTouched();
+    if (isValid) agg.write();
+  };
+
+  const onChange = (opts: EuiComboBoxOptionProps[]) => {
+    const selectedOpt: ComboBoxOption = get(opts, '0');
+
+    setValidity(!!selectedOpt);
+    setValue(selectedOpt ? selectedOpt.key : selectedOpt);
+    setTouched();
+    if (selectedOpt) agg.write();
+  };
+
+  useEffect(
+    () => {
+      setValidity(value ? (definedOption ? true : parseInterval(value, timeBase)) : false);
```

Comment:
Similar comment to my previous ones about validity. It seems that the "isValid" logic should be a computed value. It would then be much easier to follow what it is and how it is determined and how it may change over time, rather than having it imperatively set in various event handlers.

Just a thought.